org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java

```diff
@@ -549,6 +560,32 @@ private ICompilationUnit checkPackageDeclaration(String uri, ICompilationUnit un
 			} catch (CoreException e) {
 				JavaLanguageServerPlugin.logException(e.getMessage(), e);
 			}
+		} else if (!ProjectUtils.isVisibleProject(project)) {
+			PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+			List<String> sourcePaths = preferencesManager.getPreferences().getInvisibleProjectSourcePaths();
+
+			// user already set the source paths manually, we don't infer it anymore.
+			if (sourcePaths != null) {
+				return unit;
+			}
+
+			IPath projectRealFolder = ProjectUtils.getProjectRealFolder(project);
+			IPath unitPath = unit.getResource().getLocation();
+			
+			boolean needToInfer = false;
+			if (javaProject.isOnClasspath(unit)) {
+				CompilationUnit astRoot = CoreASTProvider.getInstance().getAST(unit, CoreASTProvider.WAIT_YES, new NullProgressMonitor());
+				IProblem[] problems = astRoot.getProblems();
+				needToInfer = Arrays.stream(problems).anyMatch(p -> p.getID() == IProblem.PackageIsNotExpectedPackage);
```

Comment:
One suggestion is to search multiple child files on the same folder to see if they are suffering from the same PackageIsNotExpectedPackage issue. If so, it makes sense to reinfer the source root. If some of Java files at the same level don't have the package issue, just leave the source path unchanged.

Reply:
Good idea.